### DISCUSSION TITLE
Unexpected behaviour with MediaTypeHeaderDelegate's cached values

### DISCUSSION THREAD

**Discussion by hngu-apple** [resteasy/resteasy#4230] [General]
We found an interesting issue with Media Type headers in one of our applications. Incoming requests will cache the `Accept` header value and subsequent requests may return a case-different `Content-Type` response header. I believe this is technically correct per the RFC https://www.rfc-editor.org/rfc/rfc9110.html#name-media-type but it does not seem like ideal behaviour that clients can impact the case sensitivity of response headers for other clients. 

This is the original commit that introduced the cache: https://github.com/resteasy/resteasy/commit/ea17ab76d5ec68b344ba87acbee265e8ad3e9ac1

We found this issue after upgrading from 3.6.2.Final -> 4.5.6.Final.

e.g. 
Request:
`Accept: application/vnd.api+json`
Response:
`Content-Type: application/vnd.Api+json`

Reproducible with this unit test:
```
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;

import javax.ws.rs.core.MediaType;
import org.jboss.resteasy.plugins.delegates.MediaTypeHeaderDelegate;
import org.junit.jupiter.api.Test;

public class MediaTypeHeaderDelegateTest {

    @Test
    void test() {
        final String lowerCase = "application/vnd.api+json";
        final String upperCase = "application/VND.API+JSON";

        final MediaType mediaTypeLower = MediaTypeHeaderDelegate.parse(lowerCase);
        final MediaType mediaTypeUpper = MediaTypeHeaderDelegate.parse(upperCase);

        assertThat(mediaTypeLower.toString(), is(lowerCase)); <- failing
        assertThat(mediaTypeUpper.toString(), is(upperCase));
    }

}
```

```
java.lang.AssertionError: 
Expected: is "application/vnd.api+json"
     but: was "application/VND.API+JSON"
Expected :application/vnd.api+json
Actual   :application/VND.API+JSON
```

**Comment by jamezp** [comment]
As you pointed out the media type values are meant to be case-insensitive. I don't think there is much we can do about that, especially with caching enabled.

**Comment by hngu-apple** [comment]
After doing some more investigation it looks like this was fixed https://github.com/resteasy/resteasy/commit/62ff36be5c1dd8794168ea6aefcd0bbfe24c0faf. It's available as early as 4.6.1.Final.

**Reply by jamezp** [reply]
Ah, good catch. 

For my own knowledge, is there something that keeps you on the 4.x releases?

**Reply by hngu-apple** [reply]
We are following an upgrade path outlined in https://docs.jboss.org/resteasy/docs/4.0.0.Final/userguide/html/Migration_from_older_versions.html to take us from 3.6.2 to 4.5.6. I am not sure about after this but I don't see any reason for us to remain on 4.x after we find some stability.

**Reply by jamezp** [reply]
Thank you for the insight. One thing to note support for Jakarta REST 2.1 is out of maintenance with Jakarta REST 3.1 and Jakarta REST 4.0 being the two supported API's. RESTEasy 6.2 implements Jakarta REST 3.1 and RESTEasy 7.x implements Jakarta REST 4.0.

The biggest change with these is the `javax` to `jakarta` namespace change.

**Comment by hngu-apple** [comment]
Thanks for the help James. Going to close this issue as resolved.
